ENH: Removing TransformType key from Transform serialization

PranjalSahu opened the pull request:
- Removed TransformType key from Transform dict.
- Added test for BSpline Transform for testing transformation output before and after serialization.
- Removed the first Composite transform from the dict array. However, there is one issue with this approach. The dictionary is a list of transforms present inside the composite transform and the name for composite transform cannot be transferred here.  We need to decide if that will be needed. 
- Changed parameter setting from loop to one function call. Construct a new itkOptimizerParameters object by converting NumPy array to a list.



<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

<!-- **Thanks for contributing to ITK!** -->


Sample Json for single Transform object:
```yaml
[
{'name': '', 
'parametersValueType': 'float64', 
'inputDimension': 2, 
'outputDimension': 2, 
'transformName': 'BSplineTransform', 
'inputSpaceName': '2D_VectorSpace', 
'outputSpaceName': '2D_VectorSpace', 
'parameters': array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  2.,  2.,  5.,  5.,  2.,  2.,  0.,  0.,  0.,  0.,  0.,  0.,
        5.,  5., 10., 10.,  5.,  5.,  0.,  0.,  0.,  0.,  0.,  0.,  2.,
        2.,  5.,  5.,  2.,  2.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.]), 
'fixedParameters': array([  6.        ,   6.        , -86.04166667, -86.04166667,
        85.41666667,  85.41666667,   1.        ,   0.        ,
         0.        ,   1.        ]), 
'numberOfParameters': 72, 
'numberOfFixedParameters': 10
}
]
```

Example of composite transform dict:
```yaml
[
{
'name': 'transform0', 
'parametersValueType': 'float64', 
'inputDimension': 3, 
'outputDimension': 3, 
'transformName': 'AffineTransform', 
'inputSpaceName': 'fixedSpace0', 
'outputSpaceName': 'movingSpace0', 
'parameters': array([1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.]), 
'fixedParameters': array([0., 0., 0.]), 
'numberOfParameters': 12, 
'numberOfFixedParameters': 3
}, 
{
'name': 'transform1', 
'parametersValueType': 'float64', 
'inputDimension': 3, 
'outputDimension': 3, 
'transformName': 'DisplacementFieldTransform', 
'inputSpaceName': 'fixedSpace1', 
'outputSpaceName': 'movingSpace1', 
'parameters': array([], dtype=float64),
 'fixedParameters': array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.]), 
'numberOfParameters': 0, 
'numberOfFixedParameters': 18
},
{
'name': 'transform2', 
'parametersValueType': 'float64', 
'inputDimension': 3, 
'outputDimension': 3, 
'transformName': 'Rigid3DTransform', 
'inputSpaceName': 'fixedSpace2', 
'outputSpaceName': 'movingSpace2', 
'parameters': array([1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.]), 
'fixedParameters': array([0., 0., 0.]), 
'numberOfParameters': 12, 
'numberOfFixedParameters': 3
}
]****

PranjalSahu commented:
@thewtex Do we want to give the functionality to change the **inputSpace** and **outputSpace** fields ?
For this, we will have to add string entries in the [itk::Transform](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Transform/include/itkTransform.hxx) class definition and not just in the dict_from_transform.

thewtex commented:
> add string entries in the itk::Transform class definition and not just in the dict_from_transform.

@PranjalSahu yes, please

jhlegarreta commented on Modules/Core/Transform/include/itkTransform.h:
```diff
@@ -115,6 +115,14 @@ class ITK_TEMPLATE_EXPORT Transform : public TransformBaseTemplate<TParametersVa
+
+  /* For storing the  name of OutputSpace */
+  itkSetMacro(OutputSpaceName, std::string);
+  itkGetConstReferenceMacro(OutputSpaceName, std::string);
```
Please exercise the added methods in the corresponding C++ test.

jhlegarreta replied:
Done in 5da379a. Thanks @PranjalSahu and @thewtex for the comment to ensure that the calls return what is expected.

thewtex reviewed (commented):
@PranjalSahu awesome, thank you! I like each transform name having the `name`, `inputSpace`, `outputSpace`.  We have the composite transform `inputSpace`, `outputSpace` by looking at `composite[0]['inputSpace']`, `composite[-1]['outputSpace']` -- best not to duplicate that information.

One minor change as suggested by @jhlegarreta 

```
 std::cout << transform->GetInputSpaceName() << std::endl;
```

->

```
ITK_TEST_EXPECT_EQUAL(std::string("test_inputspace"), transform->GetInputSpaceName());
```

jhlegarreta reviewed (commented):
Looks good to me, but I am deferring to @thewtex the approval.

BTW @PranjalSahu only one of the commit messages has a body, and the imperative mood should be used in the subject line. See [CONTRIBUTING.md#commit-messages](https://github.com/InsightSoftwareConsortium/ITK/blob/23da3dfa79152648f3d38050ee147df47f41ce0a/CONTRIBUTING.md#commit-messages).

PranjalSahu commented:
> Looks good to me, but I am deferring to @thewtex the approval.
> 
> BTW @PranjalSahu only one of the commit messages has a body, and the imperative mood should be used in the subject line of any commit. See [CONTRIBUTING.md#commit-messages](https://github.com/InsightSoftwareConsortium/ITK/blob/23da3dfa79152648f3d38050ee147df47f41ce0a/CONTRIBUTING.md#commit-messages).

I have added more messages and changed the commit subject.

PranjalSahu commented:
I am doing some additional testing locally with the ICON registration transform.

jhlegarreta commented:
> I have added more messages and changed the commit subject.

:100: Thanks.

PranjalSahu commented:
Please wait for my confirmation before merging.

PranjalSahu commented:
Ok, it can be merged 👍.
Tested the output with ICON registration model also.

thewtex reviewed (changes requested):
@PranjalSahu awesome!

One minor change noted inline per discussion.

thewtex commented:
CC @HastingsGreer

thewtex commented:
:100: :1st_place_medal:
